src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,224 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE 1
+
+#include "wsinstance.h"
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <errno.h>
+#include <sys/wait.h>
+#include <fcntl.h>
+#include <stdlib.h>
+
+#define HTTPS_SOCKET "https.sock"
+#define HTTP_REDIRECT_SOCKET "http_redirect.sock"
+#define HTTP_SOCKET "http.sock"
+#define SOCKET_MODES 3
+#define SD_LISTEN_FDS_START 3
+
+int socket_to_pid[SOCKET_MODES];
+struct pollfd ufds[SOCKET_MODES];
+
+static void
+handle_cockpit_ws_exited ()
+{
+  pid_t pid;
+  while (pid = waitpid (-1, NULL, WNOHANG), pid > 0)
+    {
+      for (int i = 0; i < SOCKET_MODES; i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = -1;
+            }
+        }
+    }
+}
+
+int
+spawn_cockpit_ws (char *ws_path, int mode)
+{
+  int rv = 0;
+  switch (mode)
+    {
+    case WS_INSTANCE_HTTP:
+      rv = execl (ws_path, ws_path, "--no-tls", "--port", "0", NULL);
+      break;
+
+    case WS_INSTANCE_HTTP_REDIRECT:
+      rv = execl (ws_path, ws_path, "--proxy-tls-redirect",
+                  "--no-tls", "--port", "0", NULL);
+      break;
+
+    case WS_INSTANCE_HTTPS:
+      rv = execl (ws_path, ws_path, "--for-tls-proxy", "--port", "0", NULL);
+      break;
+
+    default:
+      perror ("Invalid mode");
+      rv = -1;
+    }
+  return rv;
+}
+
+int
+main (int argc, char *argv[])
+{
+  int rv = 0;
+
+  if (argc != 3)
+    {
+      perror ("Usage: socket-activation-helper $WS_PATH $SOCKETS_DIR");
+      goto cleanup;
+    }
+
+  int listen_fd[SOCKET_MODES];
+  struct sockaddr_un addr[SOCKET_MODES];
+  char *ws_path = argv[1];
+  char *https_socket_path = NULL;
+  char *http_redirect_socket_path = NULL;
+  char *http_socket_path = NULL;
+
+  rv = asprintf (&https_socket_path, "%s/%s", argv[2], HTTPS_SOCKET);
+  if (rv == -1)
+    goto cleanup;
+  rv =
+    asprintf (&http_redirect_socket_path, "%s/%s", argv[2],
+              HTTP_REDIRECT_SOCKET);
+  if (rv == -1)
+    goto cleanup;
+  rv = asprintf (&http_socket_path, "%s/%s", argv[2], HTTP_SOCKET);
+  if (rv == -1)
+    goto cleanup;
+
+  char *socket_paths[] =
+    { https_socket_path, http_redirect_socket_path, http_socket_path };
+
+  for (int i = 0; i < SOCKET_MODES; i++)
+    {
+      socket_to_pid[i] = -1;
+      unlink (socket_paths[i]);
+
+      /* create a listening socket for each cockpit-ws mode */
+      listen_fd[i] = socket (AF_UNIX, SOCK_STREAM, 0);
+
+      memset (&addr[i], 0, sizeof (struct sockaddr_un));
+      addr[i].sun_family = AF_UNIX;
+      strcpy (addr[i].sun_path, socket_paths[i]);
+      rv = bind (listen_fd[i],
+                 (struct sockaddr *) &addr[i], sizeof (addr[i]));
```

Comment:
`sizeof` should only be used with brackets for type names or complex subexpressions.  When applied to values it's better to write `sizeof addr[i]`.